external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/trident/KafkaTridentSpoutEmitter.java

```diff
@@ -47,70 +33,111 @@
+        this.translator = kafkaManager.getKafkaSpoutConfig().getTranslator();
 
         final KafkaSpoutConfig<K, V> kafkaSpoutConfig = kafkaManager.getKafkaSpoutConfig();
-        pollTimeoutMs = kafkaSpoutConfig.getPollTimeoutMs();
-        firstPollOffsetStrategy = kafkaSpoutConfig.getFirstPollOffsetStrategy();
+        this.pollTimeoutMs = kafkaSpoutConfig.getPollTimeoutMs();
+        this.firstPollOffsetStrategy = kafkaSpoutConfig.getFirstPollOffsetStrategy();
         LOG.debug("Created {}", this);
     }
 
+    /**
+     * Creates instance of this class with default 500 millisecond refresh subscription timer
+     */
+    public KafkaTridentSpoutEmitter(KafkaTridentSpoutManager<K,V> kafkaManager, TopologyContext topologyContext) {
+        this(kafkaManager, topologyContext, new Timer(500,
+                kafkaManager.getKafkaSpoutConfig().getPartitionRefreshPeriodMs(), TimeUnit.MILLISECONDS));
+    }
+
     @Override
     public KafkaTridentSpoutBatchMetadata<K, V> emitPartitionBatch(TransactionAttempt tx, TridentCollector collector,
-            KafkaTridentSpoutTopicPartition partitionTs, KafkaTridentSpoutBatchMetadata<K, V> lastBatch) {
-        LOG.debug("Emitting batch: [transaction = {}], [partition = {}], [collector = {}], [lastBatchMetadata = {}]",
-                tx, partitionTs, collector, lastBatch);
+            KafkaTridentSpoutTopicPartition currBatchPartition, KafkaTridentSpoutBatchMetadata<K, V> lastBatch) {
 
-        final TopicPartition topicPartition = partitionTs.getTopicPartition();
+        LOG.debug("Processing batch: [transaction = {}], [currBatchPartition = {}], [lastBatchMetadata = {}], [collector = {}]",
+                tx, currBatchPartition, lastBatch, collector);
+
+        final TopicPartition currBatchTp = currBatchPartition.getTopicPartition();
+        final Set<TopicPartition> assignments = kafkaConsumer.assignment();
         KafkaTridentSpoutBatchMetadata<K, V> currentBatch = lastBatch;
         Collection<TopicPartition> pausedTopicPartitions = Collections.emptySet();
 
-        try {
-            // pause other topic partitions to only poll from current topic partition
-            pausedTopicPartitions = pauseTopicPartitions(topicPartition);
+        if (assignments == null || !assignments.contains(currBatchPartition.getTopicPartition())) {
+            LOG.debug("SKIPPING processing batch: [transaction = {}], [currBatchPartition = {}], [lastBatchMetadata = {}], " +
+                            "[collector = {}] because it is not assigned [{}] to consumer instance [{}] of consumer group [{}]",
+                    tx, currBatchPartition, lastBatch, collector, assignments, kafkaConsumer,
+                    kafkaManager.getKafkaSpoutConfig().getConsumerGroupId());
+        } else {
+            try {
+                // pause other topic partitions to only poll from current topic partition
+                pausedTopicPartitions = pauseTopicPartitions(currBatchTp);
```

Comment:
I think it will work fine. Each executor will have one instance of `KakfkaConsumer`. The org.apache.storm.trident.spout.OpaquePartitionedTridentSpoutExecutor.Emitter class has some logic that emits one batch per topic-partition. I believe the same happens in the Kafka Spout with the old consumer. In case I am misunderstanding, can you please elaborate a bit more. Thanks.